### PR TITLE
Add parsing metadata from crud response

### LINKED ISSUE #272: Parse metadata from crud response
image

### PR DESCRIPTION
**Pull request by ArtDu**
Now we can obtain metadata from the response and don't use metadata from ddl.

I didn't forget about:

- [x] Tests
- [x] Changelog
- [ ] Documentation
- [x] Clean up the code for review. See [checklist](https://github.com/tarantool/cartridge-java/blob/master/docs/review-checklist.md)

Closes #272

### CODE REVIEW

File: src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java [hunk 1]
```diff
@@ -57,16 +66,68 @@ public TarantoolResultImpl(Value value, ValueConverter<ArrayValue, T> tupleConve
         }
     }
 
+    private TarantoolSpaceMetadata parseCRUDMetadata(Value metadata) {
+        if (!metadata.isArrayValue()) {
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java:70]: Looks like potential NullPointerException, I guess will be good idea add fast fail if metadata is null

File: src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java [hunk 2]
```diff
@@ -57,16 +66,68 @@ public TarantoolResultImpl(Value value, ValueConverter<ArrayValue, T> tupleConve
...
+        for (Value value : responseFormat) {
+            if (!value.isMapValue()) {
+                throw new TarantoolClientException("Unsupported crud metadata format: expected array of maps");
+            }
+            Map<Value, Value> fieldMap = value.asMapValue().map();
+
+            Value fieldNameValue = fieldMap.get(FORMAT_NAME_KEY);
+            if (fieldNameValue == null || !fieldNameValue.isStringValue()) {
+                throw new TarantoolClientException(
+                        "Unsupported space metadata format: key '" + FORMAT_NAME_KEY + "' must have string value");
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java:87]: add current value to exception log

File: src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java [hunk 3]
```diff
@@ -57,16 +66,68 @@ public TarantoolResultImpl(Value value, ValueConverter<ArrayValue, T> tupleConve
...
+            }
+            String fieldName = fieldNameValue.asStringValue().asString();
+
+            Value fieldTypeValue = fieldMap.get(FORMAT_TYPE_KEY);
+            if (fieldTypeValue == null || !fieldTypeValue.isStringValue()) {
+                throw new TarantoolClientException(
+                        "Unsupported space metadata format: key '" + FORMAT_TYPE_KEY + "' must have string value");
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java:94]: The same

**Comment by ArtDu** [comment]
The main problem is that now we have two converters used to parse incoming tuples:
The first level parses the top level - container - List of tuples or crud response.
The second level parses one tuple at a time.

The second level always needs to know the metadata, because it creates a TarantoolTuple and passes it there. Right now it only gets the metadata from the constructor.
The conclusion is the level that parses a particular tuple must know the metadata from the level above, if it is crudResponse.

So I see 3 options:
1) Create a new fromValue method, into which we can pass custom metadata, as I did in PR - minus that we add a method to the API, although the method is not giant terrible)
2) Combine the first and second levels - consider combining two converters into one - minus, that everything is in one place parsing the container and the tuples themselves
3) Create a converter for each request of the second level - minus the memory will be painfull grown

File: src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java (hunk 2 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java]: Fixed

**Comment by akudiyar** [comment]
I think the new converter variant (3) is the because it doesn't allow to leak the implementation details through interfaces and leaves in place the ability to have different result mappers (e.g. for libs other than Tarantool/crud or non-tuple results).

The converter instantiation is a parallel problem and should be dealt with as a whole, not in this particular place.

File: src/main/java/io/tarantool/driver/mappers/MultiValueTarantoolResultMapperFactory.java [hunk 4]
```diff
@@ -82,4 +84,11 @@ public MultiValueTarantoolResultMapperFactory(MessagePackMapper messagePackMappe
                                  Class<? extends MultiValueCallResult<T, TarantoolResult<T>>> resultClass) {
         return withMultiValueResultConverter(new TarantoolResultConverter<>(valueConverter), resultClass);
     }
+
+    public CallResultMapper<TarantoolResult<T>, MultiValueCallResult<T, TarantoolResult<T>>>
+    withTarantoolTupleResultConverter(TarantoolTupleConverter valueConverter,
+                                      Class<? extends MultiValueCallResult<T, TarantoolResult<T>>> resultClass) {
+        return withMultiValueResultConverter((ValueConverter) new TarantoolTupleResultConverter<>(valueConverter),
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/mappers/MultiValueTarantoolResultMapperFactory.java:91]: Don't understand why a cast is needed here. Probably the class needs more proper generics definition or something.

File: src/main/java/io/tarantool/driver/mappers/TarantoolTupleMultiResultMapperFactory.java [hunk 5]
```diff
@@ -42,7 +42,7 @@ public TarantoolTupleMultiResultMapperFactory(MessagePackMapper messagePackMappe
     public CallResultMapper<TarantoolResult<TarantoolTuple>,
             MultiValueCallResult<TarantoolTuple, TarantoolResult<TarantoolTuple>>>
     withDefaultTupleValueConverter(MessagePackMapper messagePackMapper, TarantoolSpaceMetadata spaceMetadata) {
-        return withTarantoolResultConverter(
+        return withTarantoolTupleResultConverter(
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/mappers/TarantoolTupleMultiResultMapperFactory.java:45]: Isn't this change breaking the plain box client?

File: src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java [hunk 6]
```diff
@@ -57,19 +68,19 @@ public TarantoolResultImpl(Value value, ValueConverter<ArrayValue, T> tupleConve
...
         this.tuples = tupleArray.list().stream()
-            .map(v -> {
-                try {
-                    return tupleConverter.fromValue(v.asArrayValue());
-                } catch (MessageTypeCastException e) {
-                    throw new TarantoolTupleConversionException(v, e);
-                }
-            })
-            .collect(Collectors.toList());
+                .map(v -> {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java:73]: nit: Please set up 4 spaces as continuation identation in Idea/whatever. These unrelated space changes distract a lot.

**Review by akudiyar — CHANGES_REQUESTED** [review]
It's a better idea overall then changing the ValueConverter interface, however the changes look very rough. I'd rather not copy-paste the whole converters code in that way. Also It may be worth thinking about passing a tuple converter factory to the result converter instead of a tuple converter, that will help eliminating the fromValue method and unnecessary casts.

File: src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java [hunk 7]
```diff
@@ -57,19 +68,19 @@ public TarantoolResultImpl(Value value, ValueConverter<ArrayValue, T> tupleConve
...
+                    try {
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolResultImpl.java:74]: I guess will be better extract try/catch to method

File: src/main/java/io/tarantool/driver/core/TarantoolTupleResultImpl.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Basic TarantoolResult implementation. Supports two types of the tuple result: a MessagePack array of arrays
+ * (as in IPROTO) and a map with format {@code {"rows": [[], ...], "metadata": []}} as returned by tarantool/crud module
+ *
+ * @author Alexey Kuzin
+ */
+public class TarantoolTupleResultImpl extends TarantoolResultImpl<TarantoolTuple> {
+
+    public TarantoolTupleResultImpl(Value value, TarantoolTupleConverter tupleConverter) {
+        if (value.isArrayValue()) {
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolTupleResultImpl.java:30]: Let's try to replace multiple if/else to case for example like this
https://stackoverflow.com/questions/5579309/is-it-possible-to-use-the-instanceof-operator-in-a-switch-statement

File: src/main/java/io/tarantool/driver/core/TarantoolTupleResultImpl.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        if (metadata == null || !metadata.isArrayValue()) {
+            throw new TarantoolClientException("Unsupported crud metadata format: expected array of maps");
+        }
+
+        List<Value> responseFormat = metadata.asArrayValue().list();
+
+        Map<String, TarantoolFieldMetadata> spaceFormatMetadata = new LinkedHashMap<>();
+        int fieldPosition = 0;
+        for (Value value : responseFormat) {
+            if (!value.isMapValue()) {
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolTupleResultImpl.java:65]: Case maybe?

File: src/main/java/io/tarantool/driver/core/TarantoolTupleResultImpl.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        spaceMetadata.setSpaceFormatMetadata(spaceFormatMetadata);
+
+        return spaceMetadata;
+    }
+
+    private void setTuples(ArrayValue tupleArray, TarantoolTupleConverter tupleConverter,
+                           TarantoolSpaceMetadata responseMetadata) {
+        this.tuples = tupleArray.list().stream()
+                .map(v -> {
+                    try {
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/core/TarantoolTupleResultImpl.java:100]: Maybe extract

File: src/main/java/io/tarantool/driver/mappers/converters/value/custom/TarantoolTupleResultConverter.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+import org.msgpack.value.ArrayValue;
+import org.msgpack.value.Value;
+
+/**
+ * @author Alexey Kuzin
+ */
+public class TarantoolTupleResultConverter<V extends Value>
+        implements ValueConverter<V, TarantoolResult<TarantoolTuple>> {
+
+    private static final long serialVersionUID = 20200708L;
```
Comment (Elishtar) [inline src/main/java/io/tarantool/driver/mappers/converters/value/custom/TarantoolTupleResultConverter.java:17]: Auto generate?

File: src/main/java/io/tarantool/driver/core/metadata/parsers/CRUDResponseMetadataParser.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package io.tarantool.driver.core.metadata.parsers;
+
+import io.tarantool.driver.core.metadata.TarantoolSpaceMetadataImpl;
+import org.msgpack.value.Value;
+
+/**
+ * @author Artyom Dubinin
+ */
+public class CRUDResponseMetadataParser implements SpaceMetadataParser {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/metadata/parsers/CRUDResponseMetadataParser.java:9]: What's the purpose of this class? Currently it looks like doing nothing.

File: src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+import io.tarantool.driver.core.metadata.TarantoolSpaceMetadataImpl;
+import io.tarantool.driver.exceptions.TarantoolClientException;
+import org.msgpack.value.Value;
+
+import java.util.Map;
+
+/**
+ * @author Artyom Dubinin
+ */
+public class DDLSpaceMetadataParser implements SpaceMetadataParser {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java:13]: This class needs some unit tests and a Javadoc.

File: src/main/java/io/tarantool/driver/core/metadata/parsers/SpaceMetadataParser.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+import org.msgpack.value.impl.ImmutableStringValueImpl;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * @author Artyom Dubinin
+ */
+public interface SpaceMetadataParser {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/metadata/parsers/SpaceMetadataParser.java:19]: This class needs some unit tests and a javadoc.

File: src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public TarantoolSpaceMetadataImpl parse(Value metadata) {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java:16]: Shouldn't it take `MapValue` explicitly?

File: src/main/java/io/tarantool/driver/core/metadata/parsers/VSpaceMetadataParser.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+import io.tarantool.driver.core.metadata.TarantoolSpaceMetadataImpl;
+import io.tarantool.driver.mappers.MessagePackValueMapper;
+import org.msgpack.value.Value;
+
+import java.util.Iterator;
+
+/**
+ * @author Artyom Dubinin
+ */
+public class VSpaceMetadataParser implements SpaceMetadataParser {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/metadata/parsers/VSpaceMetadataParser.java:12]: This class needs unit tests and a Javadoc.

**Review by akudiyar — CHANGES_REQUESTED** [review]

File: src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java (hunk 15 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java:16]: If we want to have one interface for metadata parsing we need to use `Value`. Metadata can be stored in different structures

File: src/main/java/io/tarantool/driver/core/metadata/parsers/CRUDResponseMetadataParser.java (hunk 12 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/parsers/CRUDResponseMetadataParser.java]: I changed classes. You should check

File: src/main/java/io/tarantool/driver/core/metadata/parsers/SpaceMetadataParser.java (hunk 14 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/parsers/SpaceMetadataParser.java]: Fixed

File: src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java (hunk 15 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/parsers/DDLSpaceMetadataParser.java]: I changed it

File: src/main/java/io/tarantool/driver/core/metadata/parsers/VSpaceMetadataParser.java (hunk 16 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/parsers/VSpaceMetadataParser.java]: Fixed

**Comment by ArtDu** [comment]
We can split PR to "Parse metadata from crud response" 3252ba6cd60b0e09661c378263eeac65f7799109 and "Refactor metadata parsing". But I prefer to merge "Parse metadata from crud response" 3252ba6cd60b0e09661c378263eeac65f7799109 for first

File: src/main/java/io/tarantool/driver/core/metadata/TarantoolFieldMetadataImpl.java [hunk 17]
```diff
@@ -8,7 +8,7 @@
  * @author Sergey Volgin
  * @author Artyom Dubinin
  */
-class TarantoolFieldMetadataImpl implements TarantoolFieldMetadata {
+public class TarantoolFieldMetadataImpl implements TarantoolFieldMetadata {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/core/metadata/TarantoolFieldMetadataImpl.java:11]: Do we really need to make it public? It's an internal implementation class.

**Review by akudiyar — CHANGES_REQUESTED** [review]
Interestingly, that looks pretty OK without all other changes, one question is left though

File: src/main/java/io/tarantool/driver/core/metadata/TarantoolFieldMetadataImpl.java (hunk 17 above)
Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/TarantoolFieldMetadataImpl.java:11]: There are two options here:
* We store these converters(parsers in future) in `*.core.metadata` package
* Or we open this class and store classes where they should be semantically

Comment (ArtDu) [reply src/main/java/io/tarantool/driver/core/metadata/TarantoolFieldMetadataImpl.java]: First option has been picked up

**Review by akudiyar — APPROVED** [review]
LGTM

**Pull request merged**